mutation_fragment.cc

```diff
@@ -100,35 +100,44 @@ mutation_fragment::mutation_fragment(const schema& s, reader_permit permit, stat
     : _kind(kind::static_row), _data(std::make_unique<data>(std::move(permit)))
 {
     new (&_data->_static_row) static_row(std::move(r));
-    _data->_memory.reset(reader_resources::with_memory(calculate_memory_usage(s)));
+    reset_memory(s);
 }
 
 mutation_fragment::mutation_fragment(const schema& s, reader_permit permit, clustering_row&& r)
     : _kind(kind::clustering_row), _data(std::make_unique<data>(std::move(permit)))
 {
     new (&_data->_clustering_row) clustering_row(std::move(r));
-    _data->_memory.reset(reader_resources::with_memory(calculate_memory_usage(s)));
+    reset_memory(s);
 }
 
 mutation_fragment::mutation_fragment(const schema& s, reader_permit permit, range_tombstone&& r)
     : _kind(kind::range_tombstone), _data(std::make_unique<data>(std::move(permit)))
 {
     new (&_data->_range_tombstone) range_tombstone(std::move(r));
-    _data->_memory.reset(reader_resources::with_memory(calculate_memory_usage(s)));
+    reset_memory(s);
 }
 
 mutation_fragment::mutation_fragment(const schema& s, reader_permit permit, partition_start&& r)
         : _kind(kind::partition_start), _data(std::make_unique<data>(std::move(permit)))
 {
     new (&_data->_partition_start) partition_start(std::move(r));
-    _data->_memory.reset(reader_resources::with_memory(calculate_memory_usage(s)));
+    reset_memory(s);
 }
 
 mutation_fragment::mutation_fragment(const schema& s, reader_permit permit, partition_end&& r)
         : _kind(kind::partition_end), _data(std::make_unique<data>(std::move(permit)))
 {
     new (&_data->_partition_end) partition_end(std::move(r));
-    _data->_memory.reset(reader_resources::with_memory(calculate_memory_usage(s)));
+    reset_memory(s);
+}
+
+void mutation_fragment::reset_memory(const schema& s, std::optional<reader_resources> res) {
+    try {
+        _data->_memory.reset(res ? *res : reader_resources::with_memory(calculate_memory_usage(s)));
+    } catch (...) {
+        destroy_data();
+        throw;
```

Comment:
Don't understand why this is needed. Won't the destructor destroy the data?

Reply:
```c++
    [[gnu::always_inline]]
    ~mutation_fragment() {
        if (_data) {
            destroy_data();
        }
    }
```

Reply:
Apparently not, when this is called from the constructor.